src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/PostActionModel.cs

```diff
@@ -30,93 +32,105 @@ internal class PostActionModel : ConditionedConfigurationElementBase, IPostActio
-                localizations?.TryGetValue(postActionIndex++, out actionLocalizations);
-                localizedPostActions += actionLocalizations != null ? 1 : 0;
+                string? postActionId = action.ToString(nameof(Id));
 
                 Dictionary<string, string> args = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
-
                 foreach (JProperty argInfo in action.PropertiesOf("Args"))
                 {
                     args.Add(argInfo.Name, argInfo.Value.ToString());
                 }
 
-                List<ManualInstructionModel> instructionOptions = new();
+                IPostActionLocalizationModel? actionLocalizations = null;
+
+                if (postActionId != null &&
+                    localizations != null &&
+                    localizations.TryGetValue(postActionId, out actionLocalizations) &&
+                    actionLocalizations != null)
+                {
+                    localizedPostActionCount++;
+                }
 
+                List<ManualInstructionModel> localizedInstructions = new();
                 JArray? manualInstructions = action.Get<JArray>("ManualInstructions");
 
                 if (manualInstructions != null)
                 {
-                    int localizedManualInstructions = 0;
+                    int localizedInstructionCount = 0;
                     for (int i = 0; i < manualInstructions.Count; i++)
                     {
+                        string? id = manualInstructions[i].ToString("id");
                         string? text = string.Empty;
-                        if (actionLocalizations?.Instructions.TryGetValue(i, out text) ?? false)
+                        string? localizationKey = id ?? (manualInstructions.Count == 0 ? "default" : null);
+
+                        if (id != null && (actionLocalizations?.Instructions.TryGetValue(id, out text) ?? false) ||
+                            manualInstructions.Count == 1 && (actionLocalizations?.Instructions.TryGetValue("default", out text) ?? false))
                         {
-                            localizedManualInstructions++;
+                            localizedInstructionCount++;
                         }
 
                         if (string.IsNullOrEmpty(text))
                         {
                             text = manualInstructions[i].ToString("text");
                         }
 
-                        instructionOptions.Add(new ManualInstructionModel(text ?? string.Empty, manualInstructions[i].ToString("condition")));
+                        localizedInstructions.Add(new ManualInstructionModel(id, text ?? string.Empty, manualInstructions[i].ToString("condition")));
                     }
 
-                    if (actionLocalizations?.Instructions.Count > localizedManualInstructions)
+                    if (actionLocalizations?.Instructions.Count > localizedInstructionCount)
```

Comment:
we should also check if `id`s  are unique, otherwise this check might be false positive. 
The following checks should be sufficient:
- `id`s  in post actions and in localization are unique
- `id`s in localization are subset of post action `id`s

Reply:
Added checks for uniqueness of post action and manual instruction IDs. Uniqueness of localizations are already ensured while deserializing the JSON.
I used a counter here instead of checking for subsets for performance reasons. If a problem is found, only then the two sets are compared to determine the exact cause of the error.

Reply:
Re: dedup in deserializing: we are not giving error on having entries with same ID, it would be good at least do debug level on it.